parlai/agents/examples/tra.py

```diff
@@ -0,0 +1,101 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+This is an example how to extend torch ranker agent and use it for your own purpose.
+In this example, we will just use a simple bag of words model.
+"""
+from parlai.core.torch_ranker_agent import TorchRankerAgent
+import torch
+from torch import nn
+
+
+class ExampleBagOfWordsModel(nn.Module):
+    """
+    This constructs a simple bag of words model. 
+    It contains a encoder for encoding candidates and context.
+    """
+
+    def __init__(self, opt, dictionary):
+        super().__init__()
+        self.hidden_dim = opt.get('hidden_dim', 512)
+        self.dict = dictionary
+        self.encoder = nn.EmbeddingBag(len(self.dict), self.hidden_dim)
+
+    def forward_candidates(self, cand_vecs):
+        """
+        This function will encode all the candidates for a batch. It can be different cases:
+        1. inline candidate: we have inline candidates provided for each sample, it usually happens during eval / test time
```

Comment:
this list isn't comprehensive. they are actually all described in the `_build_candidates` function docstring from TRA

Reply:
....What should I write here? Stephen asks me for a docstring here,

Reply:
"""
Encodes the candidate responses.

Recall that candidates can be inline, batch, fixed, or created from the vocabulary. These candidates are built using the `_build_candidates` function in Torch Ranker Agent (add link here for description).
"""